### PR TITLE
[FIX] hr_expense: Broken access rights

### PR DESCRIPTION
**Pull request by BHanq**
The state changes right check was only done on specific method but it
wasn't check at write level. Which allowed to bypass it.

The record rule on hr_expense_user without a check on the state is in
draft allow to change data on approved expense sheets.

### CODE REVIEW

**Comment by BHanq** [comment]
Todo : Update Translation file and add tests on write

**Review by JulienAlardot — COMMENTED** [review]
Is there a ticket assigned to this?

JSYK There are heavy changes to these in the later versions

Thx  for the fix, I'll have to check up to which version it should be forwarded as I tried to restrict the rights more and more but never had the time yet to properly refactor them :)

I've pushed a fixup to your test to highlight the issue with the current state of it :)

File: addons/hr_expense/models/hr_expense.py
Comment (BHanq) [reply addons/hr_expense/models/hr_expense.py]: The issue with the current validation process is that it can be bypassed by an RPC call using the "write" method on a draft and writing the state to something else like "approved". 

So in my opinion in doesn't break the validation process. it just has to ensure that the expense goes from draft to submit and not from draft to something else (unless you have full right access).

File: addons/hr_expense/i18n/hr_expense.pot
Comment (BHanq) [reply addons/hr_expense/i18n/hr_expense.pot:1952]: I don't get what consistence it has to be. Messages around in the .pot file looks similar to mine. 

Would you prefer 

> You don't have the rights to bypass the validation process of this expense (report)

? In order to keep a consistent error message

File: addons/hr_expense/models/hr_expense.py
Comment (BHanq) [reply addons/hr_expense/models/hr_expense.py]: Thanks ! Modified :D

File: addons/hr_expense/tests/test_expenses_access_rights.py
Comment (BHanq) [reply addons/hr_expense/tests/test_expenses_access_rights.py]: Thanks ! changed

**Comment by BHanq** [comment]
Hello @JulienAlardot ! 

Thanks a lot for your quick review

> Is there a ticket assigned to this?
I just created one https://www.odoo.com/odoo/project/2745/tasks/3942827?cids=1
> JSYK There are heavy changes to these in the later versions
Yes but the access right issue is still present in master. I think it would make sense to fix it in stable until master but if you don't approve it, I can make it only in master.  
> Thx for the fix, I'll have to check up to which version it should be forwarded as I tried to restrict the rights more and more but never had the time yet to properly refactor them :)
I got that, I don't know if a refactor is needed but ensuring that the flow can't be bypassed would already be nice ahah

> I've pushed a fixup to your test to highlight the issue with the current state of it :)
Thanks ! I fixed it, I don't know why the same test in test_expense_sheet_access_rights_approve was passing.

File: addons/hr_expense/tests/test_expenses_access_rights.py
Comment (BHanq) [reply addons/hr_expense/tests/test_expenses_access_rights.py]: self.product_c doesn't work and unit_amount is required.

Comment (BHanq) [reply addons/hr_expense/tests/test_expenses_access_rights.py]: My bad didn't get it that way sorry !

**Review by JulienAlardot — CHANGES_REQUESTED** [review]
We will have to be very careful with each FW-port
It will need to be reworked in 17.0 to include the approval_state field and the 'reported' expense state

File: addons/hr_expense/models/hr_expense.py
Comment (BHanq) [reply addons/hr_expense/models/hr_expense.py:1037]: This doesn't make sense, as an normal user (which doesn't have the group_hr_expense_manager group) wouldn't be able to submit a draft expense anymore.

**Comment by JulienAlardot** [comment]
Hello, pushed a commit with some additions in the tests to highlight some parts we overlooked in this fix :)

File: addons/hr_expense/tests/test_expenses_access_rights.py
Comment (JulienAlardot) [reply addons/hr_expense/tests/test_expenses_access_rights.py:40]: up

File: addons/hr_expense/tests/test_expenses_access_rights.py
Comment (JulienAlardot) [reply addons/hr_expense/tests/test_expenses_access_rights.py:80]: up

File: addons/hr_expense/tests/test_expenses_access_rights.py
Comment (JulienAlardot) [reply addons/hr_expense/tests/test_expenses_access_rights.py:101]: up

**Review by JulienAlardot — CHANGES_REQUESTED** [review]

File: memray-odoo-bin.134996.bin
Comment (JulienAlardot) [inline memray-odoo-bin.134996.bin:1]: I think you added this by mistake :p

Comment (BHanq) [reply memray-odoo-bin.134996.bin:1]: Woops yes thanks for noticing it !

File: addons/hr_expense/models/hr_expense.py
Comment (JulienAlardot) [inline addons/hr_expense/models/hr_expense.py:1042]:
```suggestion
            elif vals['state'] == 'approve':
```

File: addons/hr_expense/models/hr_expense.py
Comment (JulienAlardot) [inline addons/hr_expense/models/hr_expense.py:1044]:
```suggestion
            elif vals['state'] == 'cancel':
```
Else resetting to draft would test it thrice

File: addons/hr_expense/tests/test_expenses_access_rights.py
Comment (JulienAlardot) [inline addons/hr_expense/tests/test_expenses_access_rights.py:81]:
```suggestion
```
If it raises, it won't change

File: addons/hr_expense/tests/test_expenses_access_rights.py
Comment (JulienAlardot) [inline addons/hr_expense/tests/test_expenses_access_rights.py:41]:
```suggestion
```
If it raises it won't change

**Review by JulienAlardot — CHANGES_REQUESTED** [review]
Last changes then it's good to me
Good job

**Comment by BHanq** [comment]
Hi @JulienAlardot, 

Each of these remarks make sense, thanks, it has been changed.

**Review by JulienAlardot — APPROVED** [review]
LGTM @antoine162

**Review by flvr-odoo — APPROVED** [review]
LGTM, however this is not stable compliant. 
I tested that it doesn't break the usual use case even if merged in stable. 
The security reinforcement will simply not work in stable.
I'm leaving the final review to @beledouxdenis since this breaks the stable policy and I'm not confident to approve it myself.

**Review by antoine162 — APPROVED** [review]
LGTM :+1: 
@robodoo r+

@flvr-odoo I think this is worth merging even though the new record rules require updating `hr_expense` on customer DBs. The restrictions in the `write` methods already provide security benefits and don't require an update. 

As far as I can see, we don't introduce any security holes through the use of the two `sudo`s:
- a `write` is performed before the one in `action_submit_sheet`, which will check access rights
- we check who the user is before the one in `reset_expense_sheets`.

I think we shouldn't automatically update `hr_expense` on all 15.0 DBs (too much risk of destroying customer configuration), but if you think we should consider it, we can contact infra about that.

**Comment by BHanq** [comment]
Thanks Antoine ! 

> I think we shouldn't automatically update `hr_expense` on all 15.0 DBs (too much risk of destroying customer configuration), but if you think we should consider it, we can contact infra about that.

I do agree, this has no major impact, but I wanted the fix to be clean.

**Pull request closed without merging**
